conn: Initial import of a light-weight network application API

authmillenon opened the pull request:
Leight-weight, but stack agnostic. I tried to keep it as close to the socket API as possible to write very simple wrappers for that in the posix module.

~~Depends on #3608.~~

OlegHahm commented:
At a first glance, it looks like sockets.

kaspar030 commented:
... with multi-stack support ...

@OlegHahm But the interface does work without dynamic memory.

kaspar030 commented:
@authmillenon I'm sorry, but this does not seem lightweight to me.

OlegHahm commented:
Can you remind me where I need dynamic memory for sockets?

kaspar030 commented:
@OlegHahm Sorry, I meant 'preallocated or dynamic memory'.

haukepetersen commented:
+1 for the general approach, but I also don't see the light-weight here. I think we can do even more simple, like a stateless, one function call for sending UDP for example.

OlegHahm commented:
btw are we talking about a layer between application layer and lower layers or between applications and the network stack in general?

kaspar030 commented:
I don't think we need multiple stacks within the interface.

- the API fixes address types to be used, making super-slim implementations impossible
- memcpy within an interface?
- while a generic interface for any connection is nice in any high-level system, RIOT-developers will know when to use tcp, udp, whatever, and can directly call the corresponding functions. No need to abstract here, better avoid the overhead and go for simplicity.

kaspar030 commented:
@authmillenon If you drop support for multiple interfaces, you basically end up with gnrc's interface, right?

kaspar030 commented:
application -> network stack

OlegHahm commented:
> while a generic interface for any connection is nice in any high-level system, RIOT-developers will know when to use tcp, udp, whatever, and can directly call the corresponding functions. No need to abstract here, better avoid the overhead and go for simplicity.

and 

> @OlegHahm application <-> network stack

IMO that's contradicting. An application developer is not necessarily a RIOT-developer and usually doesn't care about which transport layer, routing, or whatever is used below. If we want to design an interface for applications to the network stack this has to be more generic and agnostic to lower layers. If we want to design an interface from application layer protocol implementations to lower layers (transport layer or similar), your assumption is valid (I guess).

kaspar030 commented:
@authmillenon Let me be more specific when I mean "this is not lightweight".

In order to bind to an udp port, all you need is "udp_bind(port)".
So now, if I would create a "null stack", with "null_udp_bind(port)" being empty, and write a "conn_*" implementation for it, integrate it like gnrc is integrated now,
then write a function that sets up the address struct needed for "conn_bind()", then calls "conn_bind(...)", thenn calls the "null stack" conn_null_bind() to get out the port number, and then calls "null_udp_bind()" (which does nothing). Like, a minimal usage of a stack implementing the conn API, with a stub function behind.

What do you estimate is the overhead in code, memory and stack for this bind, comparing to calling "null_udp_bind(port)"?

Whatever number you say, all of it is overhead brought by "conn", and should be avoided.

kaspar030 commented:
> An application developer is not necessarily a RIOT-developer and usually doesn't care about which transport layer, routing, or whatever is used below.

Yes. Sure. He doesn't care. We all agree on that. We want our poor baby developers to be able to shout "DO!" at the RIOT node, and it does it. We want to make it even easier than Oracle does for its well-known enterprise language.

In reality, there are a lot of compromises to be made if we want a RIOT application developer to be able to do "connect me to X", not caring about whether it's TCP or not, abstracting that technicality away. I think we should summon all devs and have a big, long talk if we want RIOT to go that way. Or just provide sane, small, efficient APIs for people that *do* want to open a TCP connection.

OlegHahm commented:
> In reality, there are a lot of compromises to be made if we want a RIOT application developer to be able to do "connect me to X", not caring about whether it's TCP or not, abstracting that technicality away. I think we should summon all devs and have a big, long talk if we want RIOT to go that way. Or just provide sane, small, efficient APIs for people that do want to open a TCP connection.

You're mixing things up. An application developer wants to transmit data over lwm2m, CoAP, MQTT, whatever... It's RIOT's duty to chose and provide the proper transport/network/link layer and routing protocol implementations for that **and** provide an API to the application developer that does not require internal knowledge about the stack. There's non inherent contradiction in making this sane, small, and efficient.

People who do want to open a TCP connection are typically no application developers.

kaspar030 commented:
> You're mixing things up. An application developer wants to transmit data over lwm2m, CoAP, MQTT, whatever...

lwm2m, CoAP, MQTT are "application" when seen from the perspective of
this proposed API. Developers using libs implementing those protocols
will not use "conn_*", but "lwm2m_*", "coap_*", "mqtt_*".

Those libs will use whatever API they need to. If at that level we're
using "conn_*", we're producing bloat.

And if we re-use libs implementing those protocols, we will probably
writing wrappers from, e.g., "mqtt_open_tcp(uint32_t target_ipv4,
uint16_t port)" looking like


```
mqtt_open_tcp(uint32_t target_ipv4, uint16_t port)
{
conn_t conn;
conn_addr_ipv4_t addr;
conn_make_ipv4_addr(&addr, target_ipv4, port);
conn_open(&conn, get_default_network_stack(), AF_INET4, addr);
...
}
```

Complicated. Bloated.

OlegHahm commented:
> lwm2m, CoAP, MQTT are "application" when seen from the perspective of this proposed API. Developers using libs implementing those protocols will not use "conn_*", but "lwm2m_*", "coap_*", "mqtt_*".

Go back to my initial question and your answer. You're talking about application **layer** <-> network stack.

kaspar030 commented:
> You're talking about application **layer** <-> network stack.

I'm talking about the API proposed by this PR.

It is a stack-agnostic light-weight "socket" API, the kind of API that
is used to write libs like mqtt or coap libraries or applications that
want to use e.g., UDP or TCP.

@authmillenon Did I get that right?

OlegHahm commented:
@OlegHahm:
> btw are we talking about a layer between application layer and lower layers or between applications and the network stack in general?

but never mind

authmillenon commented:
@kaspar030 let me get your argument straight: you say it is not lightweight because it offers multi-task support / because of `conn`.

I tend to agree, that dynamic stack support isn't really necessary (and when thinking about it really hard to do with the concept I have for sockets in mind), so I guess I can make it a little more static, using macro magic instead of switch cases and inline functions. Multi-task support can also be achieved by writing a netapi adapter for a non-profit stack ;). 

The reason this API looks so much like sockets is, because it IS a pointer based socket API. This is the main reason I did not re-purposed the socket_base header. The fact alone that I don't use file descriptors or other numerical id's makes the module itself very RAM-memory efficient (about ROM we can argue, but let me rewrite it first a little), because the module does not need to maintain a (possibly huge but largely unused) socket table, as socket_base did. The other reason I did not call it socket*whatever was to avoid user confusion *we already faced* about which socket API to use. I say: If you are an inexperienced user: use POSIX sockets (or whatever library will wrap around them) because you are used to them, which will be the only API in RIOT that has "socket"  in name, when I'm done. No ambiguity possible. More experienced users still can use `conn` without using sockets, master developers even can use "bare metal" stack APIs and if even that is to fat they can hacke the hardware directly, so that they can feel proud as the 1337 haxx0rs they are ;). In general, when it comes to user APIs I really want it to be friendly (as our slogan suggests), but optional.

authmillenon commented:
> It is a stack-agnostic light-weight "socket" API, the kind of API that
is used to write libs like mqtt or coap libraries or applications that
want to use e.g., UDP or TCP.
>
> @authmillenon Did I get that right?

In general yes. It is the base for a very slim POSIX socket API, which allows then for very easy porting of external libs.

kaspar030 commented:
> @kaspar030 <https://github.com/kaspar030> let me get your argument
> straight: you say it is not lightweight because it offers multi-task
> support / because of |conn|.

It does not seem light-weight because of multi stack support and because 
it forces use of a generalized address type.

> More experienced users still can use |conn| without using sockets,
> master developers even can use "bare metal" stack APIs and if even
> that is to fat they can hacke the hardware directly, so that they
> can feel proud as> the 1337 haxx0rs they are ;).

IMHO, we have a need for a "socket" kind of API which is *not* directly 
using a stack's API, but wich is *not* adding unneccessary bloat.
Something that makes writing libraries and applications possible without 
tying to a specific network stack.

IMHO, this proposed API is too inefficient.

> In general, when it comes to user APIs I really want it to be
> friendly (as our slogan suggests), but optional.

We need to be careful with that. At some point, user-friendlyness comes 
at a price. This API is not at the highest application developer level, 
but will be used for middle-layer libraries. Once used, slimming it down 
will be hard.

OlegHahm commented:
>  In general, when it comes to user APIs I really want it to be friendly (as our slogan suggests), but optional.

An transport layer API is not a user API in my opinion.

kaspar030 commented:
> An transport layer API is not a user API in my opinion.

What is it?

OlegHahm commented:
A networking API (like netapi).

kaspar030 commented:
@OlegHahm So what does a user API do? "user_add(const char * name)"?

OlegHahm commented:
User API: something that the application developer uses (e.g. on top of CoAP)
network API: something that is used inside the network stack (e.g. between CoAP and UDP)

authmillenon commented:
> IMHO, we have a need for a "socket" kind of API which is *not* directly
using a stack's API, but wich is *not* adding unneccessary bloat.
Something that makes writing libraries and applications possible without
tying to a specific network stack.

See,  there is what I don't understand in your argument: this "something"  what you are referring to are POSIX sockets because that it what most libraries and applications use for their network connectivity. And as far as I know are sockets stack agnostic. Or are you proposing to make our own API and patch every single app and library with a huge ifdef that catches every case?

authmillenon commented:
> It does not seem light-weight because of multi stack support and because
it forces use of a generalized address type.

As I said: assuming there is only one stack, I can adapt it to be just as efficient as if you wouldn't use `conn`, just `*_conn`. How do you propose to have the addresses then? At some point you need some kind of generic format (`struct sockaddr`) if you want to stay compatible with POSIX.

kaspar030 commented:
> there is what I don't understand in your argument: this "something"  what you are referring to are POSIX sockets because that it what most libraries and applications use for their network connectivity. And as far as I know are sockets stack agnostic. Or are you proposing to make our own API and patch every single app and library with a huge ifdef that catches every case?

Again, like discussed months ago:

1. POSIX sockets are not memory friendly as pointed out several times
2. because of that, POSIX sockets are suboptimal as socket library for
an embedded system
3. because of that, we need something slimmer, more efficient, suited
for resource constrained devices
4. directly basing applications on stack specific APIs makes them tied
to that stack, which we want to avoid
5. whatever slim socket API we end up with can be used to implement
POSIX sockets on, keeping the POSIX API's inherent memory inefficiencies
to those applications that need them

kaspar030 commented:
> How do you propose to have the addresses then?

An IPv4 address has 4 bytes, IPv6 16. That's it, no need to specify more
when opening a TCP connection or sending a UDP packet.

```
tcp_t conn;
tcp_connect6(&conn, addr, port);
```

is all that is needed.

Want a generic function?

```
tcp_connect(&conn, addr, addrlen, port)
{
	if (addr_len==4) {
		return tcp_connect4(conn,port);
        } else if (addr_len==16) {
		return tcp_connect6(conn, port);
	} else {
		return - ENOTSUP;
	}
}
```

No need to carry around "address_type, address, address_len" for every
address.

And please don't make me argue against "but maybe in the future we want
to support tcp over SMS".

> At some point you need some kind of generic format (`struct sockaddr`) if you want to stay compatible with POSIX.

No, we don't, up until in the POSIX socket implementation. That
implementation can parse these structs and call whatever actual
lower-level function is necessary. It has to make the distinction
anyways, unless we carry around the extra "address meta information"
unnecessarily through half of the stack.

authmillenon commented:
As far as I see it: this is also stack agnostic ;-).

authmillenon commented:
How about now?

(depends on #3608 for IPv4 address support, now).

kaspar030 commented:
A lot better.

Some thoughts:
- I'm beginning to think that a unified IP address type might be the way to go, as it would compile out cleanly on single-stack nodes, and in dual-stack nodes, most applications would always want to support both anyways.

- do you really think it makes sense to have a unified "connection" type, as sockets provide them?
The only real advantage I see is that applications could choose, at runtime, to use, e.g., either UDP or plain IP, using the abstracted "connection" handles.
The downside is that even clearly-bound-to-one type (e.g., a http server over TCP or a coap server using UDP) have to carry around the overhead of supporting multiple types.

- I think it is not necessary to have to provide client address structs as parameter for ```accept()```. If needed at all, that should be factored out in something like "conn_getpeeraddr()"

- I think we can omit either bind or listen and go with just one function, or am I missing something?

- we should take care designing this, as it would be nice to fit whatever we come up with in a grander scheme, like, maybe we want to use handles using file IO functions (read, write, ...)

(unrelated to general interface, I'm referring to *accept():)
- we have to unify parameter ordering.
There are three "rules" which we should be consistent with:
1. when calling multiple functions or functions-within-functions, it makes sense to have arguments at fixed positions in order to reduce code size.
2. when dealing with "objects", it's nice to always have the object ptr of the object being dealt with as first parameter. That's how all object oriented languages handle it internally, mostly because of 1.)
3. as many standard C functions do it, we should put "out" parameters first.

IMHO 1.) and 2.) go together and should have preference.

Maybe we should somehow meet and discuss this in person?

authmillenon commented:
> I'm beginning to think that a unified IP address type might be the way to go, as it would compile out cleanly on single-stack nodes, and in dual-stack nodes, most applications would always want to support both anyways.

Do you mean that in terms of a replacement for an `uint8_t` array or so that the clunky `conn_.*[46]?()` functions wouldn't be neccessary? Basically do you mean option 1 or option 2 (while coding the examples I realized that 2 is a superset of 1):

```C
/* option 1 */
typedef union {
    ipv4_addr_t i4;
    ipv6_addr_t i6;
    uint8_t a8[sizeof(ipv6_addr_t];
} ip_addr1_t;

/* option 2 */
typedef enum {
    IP_ADDR_TYPE_UNDEF = 0,
    IP_ADDR_TYPE_IPV4,
    IP_ADDR_TYPE_IPV6,
} ip_addr_type_t;

typedef struct {
    ip_addr_type_t type;
    ip_addr1_t addr;
} ip_addr2_t;
```

> do you really think it makes sense to have a unified "connection" type, as sockets provide them?
The only real advantage I see is that applications could choose, at runtime, to use, e.g., either UDP or plain IP, using the abstracted "connection" handles.
The downside is that even clearly-bound-to-one type (e.g., a http server over TCP or a coap server using UDP) have to carry around the overhead of supporting multiple types.

Mh… might be worth thinking about. One could even argue to have divide connection types up for IPv4 and IPv6 connections, so that the bulk of a maximum array length for IPv6 addresses wouldn't generate overhead for pure IPv4 connections. I don't even see your advantage for a unified connection type given as an advantage: one could unionize those different types to "save" memory space (but creating overhead for more memory efficient connections) and than call the function as required by the actual connection type (which they would need to do with the current API anyway).

> I think it is not necessary to have to provide client address structs as parameter for accept(). If needed at all, that should be factored out in something like "conn_getpeeraddr()"

True.

> I think we can omit either bind or listen and go with just one function, or am I missing something?

An implementation might choose to initialize it's listen queue only at the point `listen()` is called, so I would be hesitant to merge this functionality with `bind()` (also thinking about a clear socket interface here, where the functions actually do what is specified and aren't just dummies because the underlying API does not support for it).

> we should take care designing this, as it would be nice to fit whatever we come up with in a grander scheme, like, maybe we want to use handles using file IO functions (read, write, ...)

I would put generalizing IO an abstraction layer higher as this one and keep this API network-related only.

(adapted the formatting a bit to be clearer in the next quote):
> (unrelated to general interface, I'm referring to *accept():)
> 
> * we have to unify parameter ordering. There are three "rules" which we should be consistent with:
>   1. when calling multiple functions or functions-within-functions, it makes sense to have arguments at fixed positions in order to reduce code size.
>   2. when dealing with "objects", it's nice to always have the object ptr of the object being dealt with as first parameter. That's how all object oriented languages handle it internally, mostly because of 1.)
>   3. as many standard C functions do it, we should put "out" parameters first.
>
> IMHO 1.) and 2.) go together and should have preference.

That's why I had it in the order (`conn`, `addr`, `out_conn`) in the original proposed API. Don't know anymore, why I changed it :-/ for this iteration

> Maybe we should somehow meet and discuss this in person?

Given that this post is now somewhat longish and took me over an hour to write down, I would agree with this :D Would it be possible for you to be at FU around next week?

authmillenon commented:
Now there is a connection type (and function set) for each … type of a connection:

* `conn_ip4_t`
* `conn_ip6_t`
* `conn_tcp4_t`
* `conn_tcp6_t`
* `conn_udp4_t`
* `conn_udp6_t`

Not that these types are not necessarily related to each other (at least the API does not impose this on the implementation). I decided to make this modular so that an implementation can choose to only implement a subset of connection types (e.g. if it doesn't support TCP or IPv4).

Have not adapted the GNRC example implementation yet.

authmillenon commented:
Now there is something for GNRC. :-)

authmillenon commented:
I leave the tcp6 implementation to @brummer-simon ;)

kaspar030 commented:
@authmillenon I'd be free on wednesday to iron out the last details of this API.

Thoughts:
- I think "conn_*_create" can be dropped, as either _bind() or _connect() can initialize the structure
- so why do we need both bind and listen? In 99% of all POSIX socket applications, they're used in consequtive lines, so having both seems useless to me. 

 > An implementation might choose to initialize it's listen queue only at the point listen() is called

What would be the advantage of this? Why would an application "bind" but not "listen"?

- tcp recvfrom doesn't need remote address out parameters, they can be fetched with tcp_getpeeraddress.
Also, while having "recvfrom" and "sendto" for all types is more consistent, for tcp connections I feel "read" and "write" are more fitting

authmillenon commented:
> @authmillenon I'd be free on wednesday to iron out the last details of this API.

Then I'll be there after 10:00 at least.

> * I think "conn_*_create" can be dropped, as either _bind() or _connect() can initialize the structure
> * so why do we need both bind and listen? In 99% of all POSIX socket applications, they're used in consequtive lines, so having both seems useless to me.

* For both: I like to keep the function's functionality as minimal as possible (with sockets in mind, but also to be compatible with most stuff that is out there including stacks that bring a "light-weight" socket model and lastly to keep a "best"-practice model for people used to the socket-API). I see however, that a merge of _bind and _create might make sense.
* For bind: yes and know, they are used in "consecutive" lines, but between them is in most cases a `while(1)` and after the accept either a `fork()` or `select()`. I'm not quite sure how your proposal of queue initialization + queue-get would work. Wouldn't a queue initialization reset the queue and make stacks that utilize their own queue make less efficient.

> What would be the advantage of this? Why would an application "bind" but not "listen"?

That we support these stacks that utilize accept-queues and also have a agnostic abstraction layer to "our" socket API.

> * tcp recvfrom doesn't need remote address out parameters, they can be fetched with tcp_getpeeraddress. Also, while having "recvfrom" and "sendto" for all types is more consistent, for tcp connections I feel "read" and "write" are more fitting

You are right, I was focussed on consistency, but while thinking about it, I saw that it would make more sense to implement upper-layers as a switch-case anyway.

authmillenon commented:
Merged create/bind and simplified tcp[46]_send/tcp[46]_recv.

kaspar030 commented:
> For both: I like to keep the function's functionality as minimal as possible (with sockets in mind, but also to be compatible with most stuff that is out there including stacks that bring a "light-weight" socket model and lastly to keep a "best"-practice model for people used to the socket-API). I see however, that a merge of _bind and _create might make sense.

I don't get the content of that sentence.
- "_create" is just not needed

> For bind: yes and know, they are used in "consecutive" lines, but between them is in most cases a while(1) and after the accept either a fork() or select(). I'm not quite sure how your proposal of queue initialization + queue-get would work.

"bind()" -> binds a local address/port pair to a "socket", erroring out if that pair is unavailable
"listen()" -> makes above "socket" actually accept connections, sets a backlog value

Those two don't make sense to seperate. I don't want to merge "accept()", isn't that the one usually appearing after while(1) and/or fork()

"bind_listen()" (bad name) can reserve the address/port pair, initialize a queue and activate listening, "accept" does the queue-get.

> Wouldn't a queue initialization reset the queue and make stacks that utilize their own queue make less efficient.

?

kaspar030 commented:
> >What would be the advantage of this? Why would an application "bind" but not "listen"?
> 
> That we support these stacks that utilize accept-queues and also have a agnostic abstraction layer to "our" socket API.

Why do you think merging the two functions would make the API not support stacks that utilize accept-queues?

kaspar030 commented:
> Then I'll be there after 10:00 at least.

Perfect, me too!

authmillenon commented:
Rebased to current master

authmillenon commented:
Rebased to current master

authmillenon commented:
(and #3608).

OlegHahm commented:
Without reading the whole discussion (sorry, if I missed something), one question: Let's say I want to implement CCN for RIOT, this API seemed to be of use for me at both ends. CCN wants to send something over any transport (TCP, UDP, IP, link layer, whatever...) and CCN wants to provide a transport for the application layer protocol. Is this feasible with this API? Particular, for the first part: do I need to implement the lower end for each possible transport or can I implement it once and just configure the chosen transport at compile time?

kaspar030 commented:
So you want to do e.g., udp or tcp over ccn?

authmillenon commented:
@OlegHahm you need an implementation of every transport you want to send it over, of course. Otherwise you can't send it. For your plans to use it to let CCN send stuff over X: Just have a look at my socket implementation how I dealt with multiple possible connection types. If no CONN module is present you still can use netdev to send CCN stuff out, as we talked about on Monday.

OlegHahm commented:
@kaspar030, no, I want to do, let's say, http over ccn over udp or ftp over ccn over ethernet.

OlegHahm commented:
@authmillenon, I understand this. The question is: do I have to implement an adapter to each transport layer or can I implement a generic one that just calls the send and receive function from the conn API. (As I said, CCN does not care for addresses. I receive a packet from A and put this information into some table. Later I use this information in that table to send back to A without caring about neither the transport nor the address format.)

authmillenon commented:
What you do on top of `conn` is your business ;-), which means, yes you even can make a generic adapter utilizing `conn`.

OlegHahm commented:
So, there will be some kind of `send(proto, dst, data)` function in the API?

authmillenon commented:
No.

OlegHahm commented:
Unfortunate.

authmillenon commented:
This API is just a lightweight wrapper around whatever provides connection to something else. If you want to unite this somehow you have to do it yourself or use sockets.

authmillenon commented:
> `send(proto, dst, data)`

Honest question: What is `dst`? You said yourself: CCN does not care about addresses, so what is it there? What is it for raw IPv4 transmission? What is it for UDP over IPv6? Instead of bloating up the implementation of this API to answer this questions it just gives a rudimentary conduit between the stack (or possibly on top of your CCNlite port) and whoever needs to either just use it or wants to answer this question ;-).

OlegHahm commented:
As I described above: CCN will put the address information when receiving and put in some data structure and use these information for sending, but being agnostic to the semantics and format of the addresses.

authmillenon commented:
IMHO that is something that CCN has to deal with then, since being agnostic about the semantics and format of the address isn't what you usually when writing a application or library.

OlegHahm commented:
Well, ideally one would be agnostic. An application usually doesn't really care about addresses or transport semantics. I was hoping that this API could provide something like this, but I understand now that I'll need something different.

authmillenon commented:
Any UNIX program:
```C
s = socket(AF_INET, SOCK_STREAM, 0);
```

There you defined what addresses you will use (IPv4) and what transport semantics you want to use (stream-based communication).

If you not do that your application layer protocol usually either has a default transport layer (which would require the implementing library to do the same as above) or you can configure it (which would also require the implementing library to do the same as above, just with variables).

authmillenon commented:
Rebased to current master.

kaspar030 commented:
@OlegHahm You did miss the beginning of the discussion.

Providing a truly generic "connection" API as you'd like it to have would impose the generality overhead on *all* uses of the API. Actually @authmillenon's first iteration looked somewhat like that.

authmillenon commented:
Rebased to current master and #3608.

kaspar030 commented:
pls rebase

haukepetersen commented:
@authmillenon ping

As I see it from the discussion above, this PR is ready for final review/merge as of now and we will either touch it again later or build something beside is/on top of it for more usecases. Is that correct? In that case, lets try to get this in fast (today?) as I could use it very well for the SAFEST demo :-)

authmillenon commented:
Will rebase as soon as I am in the office.

authmillenon commented:
Rebased to current #3608

authmillenon commented:
Rebased to current master

authmillenon commented:
Reworked API to only use 3 (ip, udp, tcp) modules instead of 6 (ip4, ip6, tcp4, tcp6, udp4, udp6).

cgundogan commented:
Could you rebase this please?

cgundogan commented:
oops, you did already, sorry!

cgundogan commented on Makefile.pseudomodules:
```diff
@@ -1,3 +1,8 @@
+PSEUDOMODULES += conn_ip
+PSEUDOMODULES += conn_tcp
+PSEUDOMODULES += conn_udp
+PSEUDOMODULES += ng_netif_default
```
please remove this line (:

authmillenon replied:
Done

authmillenon commented:
Removed need for UDP and raw to be connected for sending.

authmillenon commented:
Rebased to current master

cgundogan commented:
Could you provide a test application demonstrating the usage? Would make it much simpler to review this (:

authmillenon commented:
The example application in #3646 should be a sufficient test application IMHO. :-)

authmillenon commented:
Fixed some minor error checking bug

authmillenon commented:
Rebased and squashed

OlegHahm commented:
Tested witht the example in #3646. Seems to be working.

authmillenon commented:
Shall I squash then?

OlegHahm commented:
I haven't reviewed the code, but I guess it's safe to squash anyway.

authmillenon commented:
Squashed and rebased to current master.

haukepetersen commented on sys/include/net/conn/ip.h:
```diff
@@ -0,0 +1,129 @@
+#include <stdlib.h>
+
+#ifdef MODULE_GNRC_CONN_IP
+#include "net/gnrc/conn.h"
```
why this include? The generic conn interface should not care about a GNRC specific header, right?

authmillenon replied:
Because you need some magic (that I don't know) to get gcc to recognize the width of an externally defined struct.

haukepetersen commented on sys/net/gnrc/conn/gnrc_conn.c:
```diff
@@ -0,0 +1,114 @@
+
+/**
+ * @{
+ *
```
no author? Same for files below...

OlegHahm replied:
I think none (or very few) C files written by @authmillenon have an author. Since doxygen is optional for C files, I think this is fine.

haukepetersen replied:
Just for her sake I would definitely include the author - for formal aspects its fine without, but this is how you actually mark your contributions...

authmillenon replied:
Which is weird, because my template definitely defines the author-tag....

OlegHahm replied:
> but this is how you actually mark your contributions...

What about copyrights?

authmillenon replied:
@haukepetersen that I might need to change this to "Freie Universität Berlin", but I have to check my student worker contract for that again.

OlegHahm replied:
For European law it doesn't matter anyway.

haukepetersen commented:
just these two optional remarks -> ACK once Travis...

authmillenon commented:
Addressed comments and added test application for `gnrc_conn_ip` (for `gnrc_conn_udp` please continue to use the POSIX socket example). Please review the test, too.

authmillenon commented:
addressed external comment.

haukepetersen commented on tests/conn_ip/main.c:
```diff
@@ -0,0 +1,52 @@
+int main(void)
+{
+    puts("RIOT socket example application");
+    msg_init_queue(main_msg_queue, MAIN_MSG_QUEUE_SIZE);
```
main does not receive, right?

haukepetersen commented:
re-ACK when comment for test application is addressed.

authmillenon commented:
Addressed comments and squashed immediately.

authmillenon commented:
Fixed doc error and squashed immediately.

authmillenon commented:
Fixed avr-related bug in `conn` and squashed immediately.

authmillenon commented:
Defined boards with insufficient memory for test application... Now it should work ………

authmillenon commented:
(Hopefully) one last fix.
